File: solutions/clojure/robot-simulator/1/src/robot_simulator.clj
(ns robot-simulator)

(defn robot [coord bearing]
  {:coordinates coord :bearing bearing})

(def dirs
  {:west  [-1 0]
   :north [0  1]
   :south [0  -1]
   :east  [1  0]})

(def rev-dirs (clojure.set/map-invert dirs))

(defn turn-right [dir]
  (let [[x y] (dirs dir)]
    (rev-dirs [y (* -1 x)])))

(defn turn-left [dir]
  (let [[x y] (dirs dir)]
    (rev-dirs [(* -1 y) x])))

(defn advance [coord dir]
  (let [[x y] (dirs dir)]
    (-> coord
        (update :x + x)
        (update :y + y))))

(defn step [{:keys [bearing] :as robot} inst]
  (cond-> robot
    (= \L inst) (update :bearing turn-left)
    (= \R inst) (update :bearing turn-right)
    (= \A inst) (update :coordinates advance bearing)))

(defn simulate [inst robot]
  (reduce step robot inst))

File: solutions/csharp/robot-simulator/1/RobotSimulator.cs
public enum Direction
{
    North = 0,
    East = 1,
    South = 2,
    West = 3
}

public class RobotSimulator
{
    private static readonly (int X, int Y)[] Steps =
    [
        (0, 1),
        (1, 0),
        (0, -1),
        (-1, 0),
    ];

    public RobotSimulator(Direction direction, int x, int y)
    {
        this.Direction = direction;
        this.X = x;
        this.Y = y;
    }

    public Direction Direction { get; private set; }

    public int X { get; private set; }

    public int Y { get; private set; }

    public void Move(string instructions)
    {
        foreach (var instruction in instructions)
        {
            Execute(instruction);
        }
    }

    private void Execute(char instruction)
    {
        switch (instruction)
        {
            case 'L':
                Direction = Turn(Direction, -1);
                break;
            case 'R':
                Direction = Turn(Direction, 1);
                break;
            case 'A':
                Advance();
                break;
        }
    }

    private static Direction Turn(Direction direction, int steps) =>
        (Direction)(((int)direction + steps + 4) % 4);

    private void Advance()
    {
        var step = Steps[(int)Direction];

        X += step.X;
        Y += step.Y;
    }
}
